src/modelfree/density/fit_density.py

```diff
@@ -0,0 +1,317 @@
+    :param model_kwargs: (dict) parameters for the model.
+    :param max_timesteps: (int or None) the maximum number of timesteps of data to use.
+    :param data_key: (str) key into the activation file, specifying e.g. the layer.
+    :param train_opponent: (str) the opponent ID to fit the model to.
+    :param train_frac: (float) the proportion of data to use for training (remainder is test set).
+    :return A dictionary of metrics."""
+    logger.info(f"Starting density fitting, saving to {output_dir}")
+
+    # Load data
+    activations = []
+    metadata = []
+    for opponent_id, path in activation_paths.items():
+        logger.debug(f"Loaded data for {opponent_id} from {path}")
+        act, meta = _load_and_reshape_single_file(path, data_key=data_key, opponent_id=opponent_id)
+        activations.append(act)
+        metadata.append(meta)
+    activations = np.concatenate(activations)
+    metadata = pd.concat(metadata)
+    # Flatten activations (but preserve timestep)
+    activations = activations.reshape(activations.shape[0], -1)
+
+    # Sub-sample
+    np.random.shuffle(activations)
+    metadata = metadata.sample(frac=1)
+    if max_timesteps is None:
+        max_timesteps = len(metadata)
+    activations = activations[0:max_timesteps]
+    metadata = metadata[0:max_timesteps].copy()
+
+    # Split into train, validation and test
+    opponent_mask = metadata['opponent_id'] == train_opponent
+    percentage_mask = np.random.choice([True, False], size=len(metadata),
+                                       p=[train_frac, 1 - train_frac])
+
+    metadata['is_train'] = opponent_mask & percentage_mask
+    train_data = activations[metadata['is_train']]
+
+    train_opponent_validation_mask = opponent_mask & ~percentage_mask
+    train_opponent_validation_data = activations[train_opponent_validation_mask]
+
+    # Fit model and evaluate
+    model_obj = model_class(**model_kwargs)
+    model_obj.fit(train_data)
+    metadata['log_proba'] = model_obj.score_samples(activations)
+
+    metrics = {}
+    if model_class == GaussianMixture:
+        metrics = {
+            'n_components': model_kwargs.get('n_components', 1),
+            'covariance_type': model_kwargs.get('covariance_type', 'full'),
+            'train_bic': model_obj.bic(train_data),
+            'validation_bic': model_obj.bic(train_opponent_validation_data),
+            'train_log_likelihood': model_obj.score(train_data),
+            'validation_log_likelihood': model_obj.score(train_opponent_validation_data),
+        }
+        with open(os.path.join(output_dir, "metrics.json"), 'w') as f:
+            json.dump(metrics, f)
+
+    # Save activations, metadata and model weights
+    meta_path = os.path.join(output_dir, 'metadata.csv')
+    metadata.to_csv(meta_path, index=False)
+
+    # TODO: Do we need to store this? We already have activations stored elsewhere on disk.
+    activations_path = os.path.join(output_dir, 'activations.npy')
```

Comment:
Yeah, I think not; I suspect I just did this out of an abundance of making sure I had access to the activations for debugging, but since the activations_path is stored as part of the config, we should have a pretty easy time tying the two together, so I'm OK removing this

Reply:
:+1: